src/Messages/TSMessagesManager.m

```diff
@@ -281,6 +284,15 @@ - (void)handlePreKeyBundleAsync:(OWSSignalServiceProtosEnvelope *)preKeyEnvelope
 - (void)handleEnvelope:(OWSSignalServiceProtosEnvelope *)envelope plaintextData:(NSData *)plaintextData
 {
     OWSAssert([NSThread isMainThread]);
+
+    BOOL duplicateEnvelope = [self.incomingMessageFinder existsMessageWithTimestamp:envelope.timestamp
+                                                                           sourceId:envelope.source
+                                                                     sourceDeviceId:envelope.sourceDevice];
+    if (duplicateEnvelope) {
```

Comment:
Sounds good. I'd add a comment in `existsMessageWithTimestamp:...` that describes how the query will behave for older messages.